topjohnwu/Magisk
# Add AVBv2 verify and sign capability to `magiskboot`

**Note:** When signing, it modifies the image in-place without making a backup.

The gist of what it does is it replaces:
1. the embedded public key with one generated from your private key (currently hardcoded)
2. the `boot` descriptor digest using the existing salt
3. the authentication data block with the hash and signature created from your private key

To test it, run the following commands:

```bash
./build.py binary magiskboot
adb push native/out/arm64-v8a/magiskboot /data/local/tmp
adb shell
su
cd /data/local/tmp
./magiskboot avbv2 magisk_patched-23015_xxxxxxxx.img verify
./magiskboot avbv2 magisk_patched-23015_xxxxxxxx.img sign
```

To boot the signed images, the embedded `vbmeta` `flags` should be set to 0 and the `vbmeta` partition should be stock.

The has also been integrated with the app and scripts, so it detects the vbmeta partition, sets `KEEPVBMETA`, then signs the `boot` image before flashing.

`mincrypt` has been replaced with `libcrypto` (from `boringssl`), so  `bootimg.cpp` has been updated accordingly, and while it seems to work, I'm not happy with the solution for the checksum section of `repack`.

This depends on [`libavb`](https://android.googlesource.com/platform/external/avb/+/refs/tags/android-12.0.0_r12/Android.bp#224) and [`libcrypto`](https://android.googlesource.com/platform/external/boringssl/+/refs/tags/android-12.0.0_r12/Android.bp#263).

Related to #4955, supersedes #5006
